google_guest_agent/snapshot_listener.go

```diff
@@ -0,0 +1,239 @@
+	seenPostSnapshotOperationIds, _ = lru.New(128)
+)
+
+type snapshotConfig struct {
+	Timeout               int // seconds
+	ContinueOnScriptError bool
+	PreSnapshotScriptURL  string
+	PostSnapshotScriptURL string
+	Enabled               bool
+}
+
+type invalidSnapshotConfig struct {
+	msg string
+}
+
+func (e *invalidSnapshotConfig) Error() string {
+	return fmt.Sprintf("invalid config: %s", e.msg)
+}
+
+func getSnapshotConfig() (snapshotConfig, error) {
+	cfgfile := configPath
+	if runtime.GOOS == "windows" {
+		cfgfile = winConfigPath
+	}
+
+	config, err := parseConfig(cfgfile)
+	if err != nil && !os.IsNotExist(err) {
+		logger.Errorf("Error parsing config %s: %s", cfgfile, err)
+	}
+
+	var conf snapshotConfig
+	conf.Timeout = config.Section("Snapshots").Key("timeout_in_seconds").MustInt(60)
+	conf.ContinueOnScriptError = config.Section("Snapshots").Key("continue_on_script_error").MustBool(false)
+	conf.PreSnapshotScriptURL = config.Section("Snapshots").Key("pre_snapshot_script").String()
+	conf.PostSnapshotScriptURL = config.Section("Snapshots").Key("post_snapshot_script").String()
+
+	if conf.PreSnapshotScriptURL == "" && conf.PostSnapshotScriptURL == "" {
+		return conf, &invalidSnapshotConfig{"neither pre or post snapshot script has been configured"}
+	}
+
+	return conf, nil
+}
+
+func runScript(path, disks string, config snapshotConfig) (int, pb.AgentErrorCode) {
+	logger.Infof("running guest consistent snapshot script at: %s", path)
+	cmd := exec.Command(path, disks)
+	err := cmd.Start()
+	scriptsReturnCode := 0
+	agentErrorCode := pb.AgentErrorCode_NO_ERROR
+	if err != nil {
+		logger.Errorf("failed to start script at %s: %v", path, err)
+		if os.IsNotExist(err) {
+			agentErrorCode = pb.AgentErrorCode_SCRIPT_NOT_FOUND
+		} else {
+			scriptsReturnCode, agentErrorCode = 1, pb.AgentErrorCode_UNHANDLED_SCRIPT_ERROR
+		}
+
+		if config.ContinueOnScriptError {
+			agentErrorCode = pb.AgentErrorCode_NO_ERROR
+		}
+
+		return scriptsReturnCode, agentErrorCode
+	}
+
+	done := make(chan error)
+	go func() { done <- cmd.Wait() }()
+
+	// Start timer
+	timer := time.After(time.Second * time.Duration(config.Timeout))
+
+	select {
+	case <-timer:
+		logger.Errorf("timeout while running script at %s", path)
+		// Exit code of the timeout command when it times out
+		scriptsReturnCode, agentErrorCode = 124, pb.AgentErrorCode_SCRIPT_TIMED_OUT
+	case err := <-done:
+		if err != nil {
+			exiterr := err.(*exec.ExitError)
+			status := exiterr.Sys().(syscall.WaitStatus)
+			logger.Errorf("error while running script at %s, return with exit code %d", path, status.ExitStatus())
+			scriptsReturnCode, agentErrorCode = status.ExitStatus(), pb.AgentErrorCode_UNHANDLED_SCRIPT_ERROR
+		}
+	}
+
+	if config.ContinueOnScriptError {
+		agentErrorCode = pb.AgentErrorCode_NO_ERROR
+	}
+
+	return scriptsReturnCode, agentErrorCode
+}
+
+func listenForSnapshotRequests(address string, requestChan chan<- *pb.GuestMessage) {
+	for {
```

Comment:
you have an inner loop for r.Recv(), why nest the loops?

Reply:
The idea was on broken connections we should attempt to reconnect.